client/starwhale/core/job/dag.py

```diff
@@ -0,0 +1,122 @@
+from typing import Any, Set, Dict
+
+
+class _DagData(object):
+    """The internal data of DAG"""
+
+    def __init__(self) -> None:
+        self.__graph: Dict[Any, Set[Any]] = {}
+        self.__graph_reverse: Dict[Any, Set[Any]] = {}
+
+    def vertices(self) -> Set[Any]:
+        """Get the vertices list"""
+        return set(self.__graph.keys())
+
+    def add_vertex(self, vertex: Any) -> None:
+        """Add a new vertex"""
+        if vertex not in self.__graph:
+            self.__graph[vertex] = set()
+            self.__graph_reverse[vertex] = set()
+
+    def add_edge(self, v_from: Any, v_to: Any) -> None:
+        """Add an edge from one vertex to another"""
+        self.__graph[v_from].add(v_to)
+        self.__graph_reverse[v_to].add(v_from)
+
+    def remove_edge(self, v_from: Any, v_to: Any) -> None:
+        """Remove an edge from one vertex to another"""
+        self.__graph[v_from].remove(v_to)
+        self.__graph_reverse[v_to].remove(v_from)
+
+    def successors(self, vertex: Any) -> Set[Any]:
+        """Get the successors of the specified vertex"""
+        return self.__graph[vertex]
+
+    def predecessors(self, vertex: Any) -> Set[Any]:
+        """Get the predecessors of the specified vertex"""
+        return self.__graph_reverse[vertex]
+
+
+class DAG(object):
+    def __init__(self) -> None:
+        self.__data = _DagData()
+
+    def __validate_vertex(self, *vertices: Any) -> None:
```

Comment:
The double underline has some special meanings for python. In common use, users' code avoids using double underline, unless overridden built-in method. 
In those code contexts, use a single underline. 
`__validate_vertex` -> `_validate_vertex` ?